src/Symfony/Component/Security/Http/EventListener/LoginThrottlingListener.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\RateLimiter\Limiter;
+use Symfony\Component\Security\Core\Exception\SessionLockedException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Event\CheckPassportEvent;
+use Symfony\Component\Security\Http\Event\LoginSuccessEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.2
+ */
+class LoginThrottlingListener implements EventSubscriberInterface
+{
+    private $requestStack;
+    private $limiter;
+
+    public function __construct(RequestStack $requestStack, Limiter $limiter)
+    {
+        $this->requestStack = $requestStack;
+        $this->limiter = $limiter;
+    }
+
+    public function checkPassport(CheckPassportEvent $event): void
+    {
+        $passport = $event->getPassport();
+        if (!$passport->hasBadge(UserBadge::class)) {
+            return;
+        }
+
+        $request = $this->requestStack->getMasterRequest();
+        $username = $passport->getBadge(UserBadge::class)->getUserIdentifier();
+        $limiterKey = $this->createLimiterKey($username, $request);
+
+        $limiter = $this->limiter->create($limiterKey);
+        if (!$limiter->consume()) {
+            throw new SessionLockedException();
+        }
+    }
+
+    public function onSuccessfulLogin(LoginSuccessEvent $event): void
+    {
+        $limiterKey = $this->createLimiterKey($event->getAuthenticatedToken()->getUsername(), $event->getRequest());
+        $limiter = $this->limiterFactory->createLimiter($limiterKey);
+
+        $limiter->reset();
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CheckPassportEvent::class => ['checkPassport', 64],
+            LoginSuccessEvent::class => 'onSuccessfulLogin',
+        ];
+    }
+
+    private function createLimiterKey($username, Request $request): string
+    {
+        return $username.$request->getClientIp();
+    }
```

Comment:
I wonder if/how users might want to configure what goes into the limiter key 🤔 

And... this wouldn't rate limit (I don't know if it should or not) a more sophisticated attack that uses many ip addresses to try to log in as the same user.

Note: as a sanity check, Laravel does exactly this: username + ip address.

Reply:
Yeah, it seems quite common to include the IP address in the limiter key. Tbh, I'm not sure about both: this still allows enumeration attacks with many usernames. So maybe it's indeed a good idea to make this configurable (but how :thinking:)

Reply:
The main question is how would you decide on what to use as a key? Why would I use something different from the default? Attacks are the same for all websites, right? I would not make it confiugurable at first and see how it goes.